runtime/lua/vim/lsp/completion.lua

```diff
@@ -463,6 +463,188 @@ local function request(clients, bufnr, win, ctx, callback)
+      infoContext.timer:stop()
+      infoContext.timer:close()
+      infoContext.timer = nil
+    end
+    cancel_pending()
+  end
+
+  ---@return boolean, table
+  local function is_valid()
+    local cmp_info = vim.fn.complete_info({ 'selected', 'completed' })
+    return api.nvim_buf_is_valid(infoContext.bufnr)
+      and api.nvim_get_current_buf() == infoContext.bufnr
+      and vim.startswith(api.nvim_get_mode().mode, 'i')
+      and tonumber(vim.fn.pumvisible()) == 1
+      and (vim.tbl_get(cmp_info, 'completed', 'word') or '') == infoContext.word,
+      cmp_info
+  end
+
+  return function(bufnr, param, select_word)
+    cleanup()
+
+    infoContext.bufnr = bufnr
+    infoContext.word = select_word
+    local debounce_time = next_doc_debounce()
+
+    infoContext.timer = assert(vim.uv.new_timer())
+    infoContext.timer:start(
+      debounce_time,
+      0,
+      vim.schedule_wrap(function()
+        local valid, cmp_info = is_valid()
+        if not valid then
+          cleanup()
+          return
+        end
+
+        cancel_pending()
+        local client_id = vim.tbl_get(cmp_info.completed, 'user_data', 'nvim', 'lsp', 'client_id')
+        local client = client_id and lsp.get_client_by_id(client_id)
+        if not client then
+          return
+        end
+
+        local start_time = vim.uv.hrtime()
+        infoContext.last_request_time = start_time
+
+        local ok, request_id = client:request(
+          vim.lsp.protocol.Methods.completionItem_resolve,
+          param,
+          function(err, result)
+            local end_time = vim.uv.hrtime()
+            local response_time = (end_time - start_time) * ns_to_ms
+            doc_rtt_ms = doc_compute_new_average(response_time)
+
+            if err or not result or next(result) == nil then
+              if err then
+                vim.notify(err.message, vim.log.levels.WARN)
+              end
+              return
+            end
+
+            valid, cmp_info = is_valid()
+            if not valid then
+              return
+            end
+
+            local value = vim.tbl_get(result, 'documentation', 'value')
+            if not value then
+              return
+            end
+
+            local windata = api.nvim__complete_set(cmp_info.selected, {
+              info = value,
+            })
+            local kind = vim.tbl_get(result, 'documentation', 'kind')
+            update_popup_window(windata.winid, windata.bufnr, kind)
+          end,
+          bufnr
+        )
+
+        if ok and request_id then
+          infoContext.request_ids[client.id] = request_id
+        end
+      end)
+    )
+  end
+end
+
+local debounce_info_request = cmpitem_resolve()
+
+local function on_completechanged(group, bufnr)
+  api.nvim_create_autocmd('CompleteChanged', {
+    group = group,
+    buffer = bufnr,
+    callback = function(args)
+      local completed_item = api.nvim_get_vvar('event').completed_item or {}
```

Comment:
`vim.v.event` ...

why use the more verbose api? same for any other code in this PR, and future PRs...

Reply:
because this file elsewhere uses api.nvim_get_vvar('event')